compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.h

```diff
@@ -19,14 +19,109 @@ namespace iree_compiler {
 // won't get tiled if the case is matmul + generic op. In this case, we have to
 // tile along reduction dim again, which needs them to be TilingInterface ops.
-enum class TilingLevel : unsigned {
-  // Tile TilingInterface operations to threads.
-  WorkGroupTiles = 0,
-  // Tile TilingInterface operation on workgroup thread for parallel dims.
-  ParallelTiles = 1,
-  // Tile TilingInterface operations on workgroup thread for reduction dims.
-  ReductionTiles = 2,
-  NumTileLevels = 3
+// TODO: Update doc.
+class TilingConfig {
+ public:
+  TilingConfig(IREE::Codegen::LoweringConfigAttr lc) : loweringConfig(lc) {
+    int numTileLevels = loweringConfig.getTileSizes().size();
+
+    // Initialize indices as invalid.
+    int invalidIdx = std::numeric_limits<unsigned>::max();
+    for (int i = 0; i < MaxNumTileLevels; ++i) {
+      levelToIdxMap[i] = invalidIdx;
+    }
+
+    switch (numTileLevels) {
+      case 3:
+        levelToIdxMap[ReductionVectorTiles] = 2;
+        // Fall through.
+      case 2:
+        levelToIdxMap[ParallelVectorTiles] = 1;
+        // Fall through.
+      case 1:
+        levelToIdxMap[DistributionTiles] = 0;
+        break;
+      case MaxNumTileLevels:
+        for (int i = 0; i < MaxNumTileLevels; ++i) {
+          levelToIdxMap[i] = i;
+        }
+        break;
+      default:
+        llvm_unreachable("Unexpected number of tiling levels");
+    };
+  };
+
+  unsigned getNumLevels() { return loweringConfig.getTileSizes().size(); };
+
+  TileSizesListType getAllTileSizes() {
+    return loweringConfig.getTileSizeVals();
+  }
+
+  SmallVector<int64_t> getDistributionVectorSizes() {
+    return loweringConfig.getTileSizeVals(getIdx(DistributionTiles));
+  }
+
+  unsigned getParallelCacheIdx() { return getIdx(ParallelCacheTiles); }
+
+  unsigned getParallelVectorIdx() { return getIdx(ParallelVectorTiles); }
+
+  unsigned getNumParallelVectorTiles() {
+    return llvm::count_if(getParallelVectorSizes(),
+                          [](int64_t tileSize) { return tileSize != 0; });
+  }
+
+  SmallVector<int64_t> getParallelVectorSizes() {
+    return loweringConfig.getTileSizeVals(getParallelVectorIdx());
+  }
+
+  unsigned getReductionCacheIdx() { return getIdx(ReductionCacheTiles); }
+
+  unsigned getReductionVectorIdx() { return getIdx(ReductionVectorTiles); }
+
+  SmallVector<int64_t> getReductionVectorSizes() {
+    return loweringConfig.getTileSizeVals(getReductionVectorIdx());
+  }
+
+  SmallVector<int64_t> getFusableLevels() {
+    switch (getNumLevels()) {
+      case 0:
+        return {};
+      case 1:
+        // Only distribution level.
+        return {0};
+      case 3:
+        // Distribution + parallel vectorization levels.
+        return {0, 1};
+      case 5:
+        // Distribution + parallel cache and vectorization levels.
+        return {0, 1, 2};
+      default:
+        llvm_unreachable("Unexpected number of tiling levels");
+    }
+  }
+
+ private:
+  enum TilingLevel {
```

Comment:
can we use enum class? It can usually reduce namespace pollution and avoid bugs in implicit conversions.

I learnt it from https://abseil.io/tips/86

Reply:
I think it is fine in this case because it's scoped within class private domain.